Source/WebKit/WebProcess/InjectedBundle/API/glib/WebKitWebExtension.cpp

```diff
@@ -172,6 +210,27 @@ static void webkit_web_extension_class_init(WebKitWebExtensionClass* klass)
         g_cclosure_marshal_generic,
         G_TYPE_NONE, 1,
         WEBKIT_TYPE_USER_MESSAGE);
+
+    /**
+     * WebKitWebExtension::initialize-web-extensions:
```

Comment:
So my problem with this pull request is that, with no way to associate the message with any particular web page, you didn't actually accomplish the original goal, right? I would add a  `page-created-with-user-data` signal instead, which would solve this problem better than `initialize-web-extensions`. (Then in the future upcoming API version, we can get rid of the original `page-created` and rename the new `page-created-with-user-data` to `page-created`.)

Reply:
That is an interesting idea. If I understood well, I could implement that by modifying the CreateWebPage message by adding the `getInjectedBundleInitializationUserDataByPage(*this, page)` userData in the message:

```diff
- Source/WebKit/WebProcess/WebProcess.messages.in:    CreateWebPage(WebCore::PageIdentifier newPageID, struct WebKit::WebPageCreationParameters pageCreationParameters)
+ Source/WebKit/WebProcess/WebProcess.messages.in:    CreateWebPage(WebCore::PageIdentifier newPageID, struct WebKit::WebPageCreationParameters pageCreationParameters, WebKit::UserData userData)
```
Then, the message is used in this 2 places. Not sure if I can get access to the `std::unique_ptr<WebPageInjectedBundleClient> m_injectedBundleClient;` in there:

```
Source/WebKit/UIProcess/WebPageProxy.cpp:    send(Messages::WebProcess::CreateWebPage(m_webPageID, creationParameters(m_process, *m_drawingArea)), 0);
Source/WebKit/UIProcess/ProvisionalPageProxy.cpp:    m_process->send(Messages::WebProcess::CreateWebPage(m_webPageID, parameters), 0);
```

Reply:
Notice that I need access to the WebPageInjectedBundleClient because it  is the one who has the reference to the WebContext. A reference that I need to get the WebView using  `webkitWebContextGetWebViewForPage(m_webContext, &page)`

Reply:
> Not sure if I can get access to the std::unique_ptr<WebPageInjectedBundleClient> m_injectedBundleClient; in there:

Hmm...

> better thought, I dont think changing the signature of `CreateWebPage(WebCore::PageIdentifier newPageID, struct WebKit::WebPageCreationParameters pageCreationParameters) could be a good idea. That actually breaks the current API, right?

This is an internal IPC format, not an API, so feel free to change it however you please. However, ideally you would just add the new parameter to WebPageCreationParameters. You'll have to modify the serialization and deserialization of WebPageCreationParameters, but that's not difficult.

Reply:
Just to ensure I catch you. You are suggesting to move the user data as part of the WebView properties and pass it as part of the `pageConfiguration` in the `webkitWebContextCreatePageForWebView()`. is that?. Something like this? 



``` diff
diff --git a/Source/WebKit/UIProcess/API/glib/WebKitWebContext.cpp b/Source/WebKit/UIProcess/API/glib/WebKitWebContext.cpp
index 1b128a4c..3d1c761e 100644
--- a/Source/WebKit/UIProcess/API/glib/WebKitWebContext.cpp
+++ b/Source/WebKit/UIProcess/API/glib/WebKitWebContext.cpp
@@ -1893,6 +1893,7 @@ void webkitWebContextCreatePageForWebView(WebKitWebContext* context, WebKitWebVi
     pageConfiguration->setRelatedPage(relatedView ? &webkitWebViewGetPage(relatedView) : nullptr);
     pageConfiguration->setUserContentController(userContentManager ? webkitUserContentManagerGetUserContentControllerProxy(userContentManager) : nullptr);
     pageConfiguration->setControlledByAutomation(webkit_web_view_is_controlled_by_automation(webView));
+    pageConfiguration->setUseData(webkit_web_view_get_user_data(webView) : nullptr);
 
     WebKitWebsiteDataManager* manager = webkitWebViewGetWebsiteDataManager(webView);
     if (!manager)
diff --git a/Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp b/Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp
index 845dda08..28c7838f 100644
--- a/Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp
+++ b/Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp
@@ -213,6 +213,8 @@ enum {
     PROP_MICROPHONE_CAPTURE_STATE,
     PROP_DISPLAY_CAPTURE_STATE,
 
+    PROP_USER_DATA,
+
     N_PROPERTIES,
 };
 
@@ -905,6 +907,10 @@ static void webkitWebViewSetProperty(GObject* object, guint propId, const GValue
     case PROP_DISPLAY_CAPTURE_STATE:
         webkit_web_view_set_display_capture_state(webView, static_cast<WebKitMediaCaptureState>(g_value_get_enum(value)));
         break;
+    case PROP_USER_DATA:
+	gpointer userData = g_value_get_object(value);
+        webView->priv->userData = userData ? G_TYPE_CHECK_INSTANCE_CAST(userData, G_TYPE_VARIANT, GVariant) : nullptr;
+        break;
     default:
         G_OBJECT_WARN_INVALID_PROPERTY_ID(object, propId, paramSpec);
     }
@@ -985,6 +991,9 @@ static void webkitWebViewGetProperty(GObject* object, guint propId, GValue* valu
     case PROP_DISPLAY_CAPTURE_STATE:
         g_value_set_enum(value, webkit_web_view_get_display_capture_state(webView));
         break;
+    case PROP_USER_DATA:
+        g_value_set_variant(value, webView->priv->userData.get());
+        break;
     default:
         G_OBJECT_WARN_INVALID_PROPERTY_ID(object, propId, paramSpec);
     }
@@ -1454,6 +1463,16 @@ static void webkit_web_view_class_init(WebKitWebViewClass* webViewClass)
         WEBKIT_MEDIA_CAPTURE_STATE_NONE,
         WEBKIT_PARAM_READWRITE);
 
+    /*
+     * Since: 2.38
+     */
+    sObjProperties[PROP_USER_DATA] = g_param_spec_object(
+        "user-data",
+        "User Data",
+        _("Associated custom user data"),
+        G_TYPE_VARIANT,
+        WEBKIT_PARAM_READWRITE);
+
     g_object_class_install_properties(gObjectClass, N_PROPERTIES, sObjProperties);
 
     /**

```

Then add it to the `WebPageCreationParameters` attached to the `CreateWebPage` IPC message and expose it as userData in a `page-created-user-data` signal.

Reply:
> just add a WebKitWebView::page-created signal

Also, the application can use the existing IPC APIs and you don't need to worry about it. All you'd need to do is make sure the signal is emitted at the right time, and the application can handle the rest.

Reply:
You could get close to a `WebKitWebView::page-created` by firing a signal from `PageClient::didChangeWebPageID`. That's called from `WebPageProxy::swapToProvisionalPage.` Swap to provisional page is not the same thing as page creation, though....